### ISSUE TITLE
README instructions for analyzing xen hypervisor are incomplete

### ISSUE THREAD

**Issue by Jakob-Koschel** [spectector/spectector-benchmarks#1]
I'm trying to reproduce the results for analyzing the xen hypervisor using spectector.

I did the instructed steps:
1. git clone `https://xenbits.xen.org/git-http/xen.git` (revision: `b541287c3600713feaaaf7608cd405e7b2e4efd0`)
2. run ./configure (output attached)
3. modify the `xen/Makefile` (git diff attached)
4. run `./obtain_project_files.sh ~/xen/xen target/xen_no_linked` (output attached)

It seems like there are no rules setup to generate `*.ll` files so there should be more steps involved to generate those.

Expanding the README would be appreciated,
Thanks!

```
./configure                                                                                                                                                                  [146/242]
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
Will build the following subsystems:
  xen
  tools
  stubdom
  docs
configure: creating ./config.status
config.status: creating config/Toplevel.mk
config.status: creating config/Paths.mk
=== configuring in tools (/home/jakob/xen/tools)
configure: running /bin/bash ./configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=.
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for as86... /usr/bin/as86
checking for ld86... /usr/bin/ld86
checking for bcc... /usr/bin/bcc
checking for lzma_version_number in -llzma... yes
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether make sets $(MAKE)... yes
checking for a BSD-compatible install... /usr/bin/install -c
checking for bison... /usr/bin/bison
checking for flex... /usr/bin/flex
checking for perl... /usr/bin/perl
checking for awk... /usr/bin/awk
checking for ocamlc... no
checking for ocaml... no
checking for ocamldep... no
checking for ocamlmktop... no
checking for ocamlmklib... no
checking for ocamldoc... no
checking for ocamlbuild... no
checking for ocamlfind... no
checking for gawk... /usr/bin/awk
checking for checkpolicy... no
checking for bash... /bin/bash
checking for python... /usr/bin/python
checking for python version >= 2.6 ... yes
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes                                                                                                                                                                     [89/242]
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for python-config... /usr/bin/python-config
checking Python.h usability... yes
checking Python.h presence... yes
checking for Python.h... yes
checking for PyArg_ParseTuple... yes
checking whether Python setup.py brokenly enables -D_FORTIFY_SOURCE... yes
checking for xgettext... /usr/bin/xgettext
checking for iasl... /usr/bin/iasl
checking uuid/uuid.h usability... yes
checking uuid/uuid.h presence... yes
checking for uuid/uuid.h... yes
checking for uuid_clear in -luuid... yes
checking uuid.h usability... no
checking uuid.h presence... no
checking for uuid.h... no
checking curses.h usability... yes
checking curses.h presence... yes
checking for curses.h... yes
checking for clear in -lcurses... yes
checking ncurses.h usability... yes
checking ncurses.h presence... yes
checking for ncurses.h... yes
checking for clear in -lncurses... yes
checking for define_key in -ltinfo... yes
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for glib... yes
checking for pixman... yes
checking for wget... /usr/bin/wget
checking for false... /bin/false
checking bzlib.h usability... no
checking bzlib.h presence... no
checking for bzlib.h... no
checking lzma.h usability... yes
checking lzma.h presence... yes
checking for lzma.h... yes
checking for lzma_stream_decoder in -llzma... yes
checking lzo/lzo1x.h usability... no
checking lzo/lzo1x.h presence... no
checking for lzo/lzo1x.h... no
checking ext2fs/ext2fs.h usability... no
checking ext2fs/ext2fs.h presence... no
checking for ext2fs/ext2fs.h... no
checking ext4fs/ext2fs.h usability... no
checking ext4fs/ext2fs.h presence... no
checking for ext4fs/ext2fs.h... no
checking for pthread flag... -pthread
checking libutil.h usability... no
checking libutil.h presence... no
checking for libutil.h... no
checking for openpty et al... -lutil
checking for yajl_alloc in -lyajl... yes
checking for deflateCopy in -lz... yes
checking for libiconv_open in -liconv... no
checking argp.h usability... yes
checking argp.h presence... yes
checking for argp.h... yes
checking for argp_usage in -largp... no
checking yajl/yajl_version.h usability... yes
checking yajl/yajl_version.h presence... yes
checking for yajl/yajl_version.h... yes
checking sys/eventfd.h usability... yes
checking sys/eventfd.h presence... yes
checking for sys/eventfd.h... yes
checking valgrind/memcheck.h usability... no
checking valgrind/memcheck.h presence... no
checking for valgrind/memcheck.h... no
checking utmp.h usability... yes
checking utmp.h presence... yes
checking for utmp.h... yes
checking for LIBNL3... no
configure: WARNING: Disabling support for Remus network buffering and COLO.
    Please install libnl3 libraries (including libnl3-route), command line tools and devel
    headers - version 3.2.8 or higher
checking for SYSTEMD... no
checking for SYSTEMD... no
configure: creating ./config.status
config.status: creating ../config/Tools.mk
config.status: creating hotplug/FreeBSD/rc.d/xencommons
config.status: creating hotplug/FreeBSD/rc.d/xendriverdomain
config.status: creating hotplug/Linux/init.d/sysconfig.xencommons
config.status: creating hotplug/Linux/init.d/sysconfig.xendomains
config.status: creating hotplug/Linux/init.d/xen-watchdog
config.status: creating hotplug/Linux/init.d/xencommons
config.status: creating hotplug/Linux/init.d/xendomains
config.status: creating hotplug/Linux/init.d/xendriverdomain
config.status: creating hotplug/Linux/launch-xenstore
config.status: creating hotplug/Linux/vif-setup
config.status: creating hotplug/Linux/xen-hotplug-common.sh
config.status: creating hotplug/Linux/xendomains
config.status: creating hotplug/NetBSD/rc.d/xencommons
config.status: creating hotplug/NetBSD/rc.d/xendriverdomain
config.status: creating ocaml/xenstored/oxenstored.conf
config.status: creating config.h
config.status: config.h is unchanged
=== configuring in stubdom (/home/jakob/xen/stubdom)
configure: running /bin/bash ./configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=.
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
xyy
checking for wget... /usr/bin/wget
checking for false... /bin/false
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether make sets $(MAKE)... yes
checking for a BSD-compatible install... /usr/bin/install -c
checking for cmake... /usr/bin/cmake
Will build the following stub domains:
  pv-grub
  xenstore-stubdom
  vtpm-stubdom
  vtpmmgr-stubdom
  ioemu-stubdom
configure: creating ./config.status
config.status: creating ../config/Stubdom.mk
=== configuring in docs (/home/jakob/xen/docs)
configure: running /bin/bash ./configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=.
checking for fig2dev... no
configure: WARNING: fig2dev is not available so some documentation won't be built
checking for pod2man... /usr/bin/pod2man
checking for pod2html... /usr/bin/pod2html
checking for pod2text... /usr/bin/pod2text
checking for pandoc... no
configure: WARNING: pandoc is not available so some documentation won't be built
checking for perl... /usr/bin/perl
configure: creating ./config.status
config.status: creating ../config/Docs.mk
config.status: creating man/xl.cfg.5.pod
config.status: creating man/xl.1.pod
```

changes to Makefile:
```
diff --git a/xen/Makefile b/xen/Makefile
index c80914c31d..87150f01c5 100644
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -16,6 +16,8 @@ export XEN_CONFIG_EXPERT ?= n
 export BASEDIR := $(CURDIR)
 export XEN_ROOT := $(BASEDIR)/..

+clang=y
+
 # Do not use make's built-in rules and variables
 MAKEFLAGS += -rR

@@ -242,7 +244,7 @@ _MAP:
 .PHONY: FORCE
 FORCE:

-%.o %.i %.s: %.c FORCE
+%.o %.i %.s %.ll: %.c FORCE
        $(MAKE) -f $(BASEDIR)/Rules.mk -C $(*D) $(@F)

 %.o %.s: %.S FORCE
```

output of `obtain_project_files.sh`:
```
Usage: obtain_project_files.sh -i PROJECT_FOLDER -o OUTDIR [clean]
make ./arch/arm/acpi/boot.ll
make -f /home/jakob/xen/xen/Rules.mk -C arch/arm/acpi boot.ll
make[1]: Entering directory '/home/jakob/xen/xen/arch/arm/acpi'
make[1]: *** No rule to make target 'boot.ll'.  Stop.
make[1]: Leaving directory '/home/jakob/xen/xen/arch/arm/acpi'
Makefile:248: recipe for target 'arch/arm/acpi/boot.ll' failed
make: *** [arch/arm/acpi/boot.ll] Error 2
make ./arch/arm/acpi/domain_build.ll
make -f /home/jakob/xen/xen/Rules.mk -C arch/arm/acpi domain_build.ll
make[1]: Entering directory '/home/jakob/xen/xen/arch/arm/acpi'
make[1]: *** No rule to make target 'domain_build.ll'.  Stop.
make[1]: Leaving directory '/home/jakob/xen/xen/arch/arm/acpi'
Makefile:248: recipe for target 'arch/arm/acpi/domain_build.ll' failed
make: *** [arch/arm/acpi/domain_build.ll] Error 2
make ./arch/arm/acpi/lib.ll
...
```

**Comment by andsanmar** [comment]
Do you've the `clang` compiler installed? If not it won't produce the `.ll` files.

**Comment by Jakob-Koschel** [comment]
Sure:

```
clang --version
clang version 7.0.0-3~ubuntu0.18.04.1 (tags/RELEASE_700/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /home/jakob/llvmlinux/llvm-bin-7/bin

llc --version
LLVM (http://llvm.org/):
  LLVM version 7.0.0

  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: znver1

  Registered Targets:
    ...
    x86        - 32-bit X86: Pentium-Pro and above
    x86-64     - 64-bit X86: EM64T and AMD64
```

**Comment by andsanmar** [comment]
Have you tried to do the same steps with the revision the README points to (`fa23f2aaa24c603f748b49b32378b738d18cc68f`)?

**Comment by Jakob-Koschel** [comment]
The README does not refer to any revision to be used for xen.
Regardless I just tried, same issue.

**Comment by andsanmar** [comment]
Add these lines to `Rules.mk` file (the one in the same directory than the `Makefile` you've modified):

```
%.ll: %.c Makefile
	$(CC) $(CFLAGS) -flto -fshort-wchar -S -emit-llvm $< -o $@
```

Try with it, and if you don't get the files compiled with clang, modify in `obtain_project_files.sh` the line:
`make $filename` (L39) by `make clang=y $filename`

Let me know if it works for you or not. Anyway, the README file is going to be updated.

**Comment by mguarnieri** [comment]
Hi Jakob,

Thanks for pointing out the issue. We updated the instruction with the missing bits.

NOTE: This repository was not synchronized with our internal repo, which contained a more recent version of benchmarks and docs (which, however, still lacked the bits about generating *.ll files). 
I took this chance to update everything to the most recent version. The new README should be easier to read (and more complete :-)).
